File: algo/63. Unique Paths II.js
// Dynamic Programming
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
    var m = obstacleGrid.length;
    var n = obstacleGrid[0].length;
    var helper_matrix = [];
    var dummy_row = [];
    for (var i=0; i<n; i++) {
        dummy_row.push(0);
    }
    for (var i=0; i<m; i++) {
        helper_matrix.push(Array.from(dummy_row));
    }
    for (var i=0; i<m; i++) {
        for (var j=0; j<n; j++) {
            if (obstacleGrid[i][j] == 1) {
                helper_matrix[i][j] = 0;
            } else if (i-1>=0 && j-1>=0) {
                helper_matrix[i][j] = helper_matrix[i-1][j] + helper_matrix[i][j-1];
            } else if (i-1>=0) {
                helper_matrix[i][j] = helper_matrix[i-1][j]
            } else if (j-1>=0) {
                helper_matrix[i][j] = helper_matrix[i][j-1]
            } else {
                helper_matrix[i][j] = 1
            }
        }
    }
    return helper_matrix[m-1][n-1];
};

File: round3/63. Unique Paths II.cs
public class Solution {
    public int UniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.Length, n = obstacleGrid[0].Length;
        var dp = new int[n];
        
        for (int j = 0; j < n; j ++) {
            if (obstacleGrid[0][j] == 1) {
                while (j < n) {
                    dp[j] = 0;
                    j ++;
                }
            } else {
                dp[j] = 1;
            }
        }
        
        for (int i = 1; i < m; i++) {
            if (obstacleGrid[i][0] == 1) dp[0] = 0;
            for (int j = 1; j < n; j++) {
                if (obstacleGrid[i][j] == 1) {
                    dp[j] = 0;
                } else {
                    dp[j] = dp[j-1] + dp[j];
                }
            }
        }
        
        return dp[n-1];
    }
}